Package libraries Java code as Java Modules

JaroslavTulach opened the pull request:
### Pull Request Description

Implements #7082 by creating a module layer per library. Removes the obsolete [add_to_class_path](https://github.com/enso-org/enso/pull/10714/commits/fb272ddcc053fb9ebfa28dd8a1497e65e2ca4358) concept as there is no longer a single `ClassPath` to load from. Each library with `polyglot/java` JARs gets its own. Need a way to express dependencies between Java APIs among multiple libraries - created `requires:` flag in `package.yaml`.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] Report [decent exception of layer resolution](https://github.com/enso-org/enso/pull/10714#issuecomment-2391006529) problem

JaroslavTulach committed:
Must specify --repl to enable debug server

JaroslavTulach committed:
Use --repl to enable REPL changelog note

JaroslavTulach committed:
Towards a ClassPath per Package

JaroslavTulach committed:
One ClassLoader per each Package

JaroslavTulach committed:
Define ModuleLayer for each Package

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java:
```diff
@@ -428,22 +432,38 @@ public Optional<Module> findModuleByExpressionId(UUID expressionId) {
+
+      this.cntrl = ModuleLayer.defineModulesWithOneLoader(cfg, parentModules, parentLoader);
+      this.layer = cntrl.layer();
+      System.out.println("modules: " + layer.modules());
```
Whatever I do, I am getting empty `[]` here. Probably I should specify something else than `emptyList()` in the `cfg` definition. But I don't know any module name to ask for! Ideas, @Akirathan?

JaroslavTulach replied:
Using `ModuleFinder.findAll` per Pavel's suggestion in 8700cc8141

JaroslavTulach commented:
Question for @radeusgd - we seem to get the encapsulation with all the consequences. There is dozen of failures:
- Incorrect polyglot import [Stream_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1118)
- Incorrect polyglot import [File_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1162)
- [CompareException](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1189)
- [Text_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1228)
- [CaseFoldedString](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1238)
- some [Not_Logged_in.Error](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1296)
- `NoClassDefFoundError` [Polyglot_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1623)
- [Autodetect read](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1680) doesn't seem to work
- Another [Polyglot_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1732)

I can think of ways to fix the tests, but some failures show _intended cross-library usage_ we don't have any alternative yet.

JaroslavTulach committed:
Load classes via ModuleLayer

radeusgd commented:
> Question for @radeusgd - we seem to get the encapsulation with all the consequences. There is dozen of failures:
> 
> * Incorrect polyglot import [Stream_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1118)
> * Incorrect polyglot import [File_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1162)
> * [CompareException](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1189)
> * [Text_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1228)
> * [CaseFoldedString](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1238)
> * some [Not_Logged_in.Error](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1296)
> * `NoClassDefFoundError` [Polyglot_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1623)
> * [Autodetect read](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1680) doesn't seem to work
> * Another [Polyglot_Utils](https://github.com/enso-org/enso/actions/runs/10158427493/job/28090470750?pr=10714#step:7:1732)
> 
> I can think of ways to fix the tests, but some failures show _intended cross-library usage_ we don't have any alternative yet.

I imagine we can come up with some workarounds, although that may not be trivial. I'm worried the most about the read autodetection - does it look like the classpath separation breaks the SPI? How can we fix that?

Alternatively, as suggested in #7082 (title _Re-exporting_), could we provide a way for a library to opt-in to being able to access modules of another lib? e.g. `Standard.Table` asking to be able to user Java helpers of `Standard.Base`, e.g. by adding some entry in its `package.yaml`? That would still prevent users from accessing the internals (unless they also add such an override) while allowing us to more easily share utilities between our libraries.

JaroslavTulach committed:
Turning std-base and std-aws into Java modules

JaroslavTulach commented on std-bits/base/src/main/java/module-info.java:
```diff
@@ -0,0 +1,14 @@
+
+module org.enso.base {
+  requires java.logging;
+  requires org.graalvm.collections;
```
@Akirathan, after 74f81117db0f6ec13f7b11e4493657983330045f I am getting:
```
[error] /home/devel/NetBeansProjects/enso/enso/std-bits/base/src/main/java/module-info.java:6:1:  error: module not found: org.graalvm.collections
[error]   requires org.graalvm.collections;
[error]                       ^1 error
[warn] frgaal exited with exit code 1
```
because the `~/.cache/coursier/v1/https/repo1.maven.org/maven2/org/graalvm/sdk/collections/24.0.0/collections-24.0.0.jar` module is on _classpath_, not _modulepath_. What do I need to change to put the module on module path?

JaroslavTulach commented:
> I can think of ways to fix the tests

My plan was:
```diff
enso$ git diff distribution/ test/Base_Tests/src/Data/Text/Utils_Spec.enso
diff --git distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Extra_Imports.enso distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Extra_Imports.enso
index 7771401cf4..c42ee36ec6 100644
--- distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Extra_Imports.enso
+++ distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Extra_Imports.enso
@@ -9,7 +9,7 @@ private
 polyglot java import java.io.PrintStream
 
 # needed by Util_Spec
-polyglot java import org.enso.base.text.CaseFoldedString
+polyglot java import org.enso.base.text.CaseFoldedString as JCaseFoldedString
 polyglot java import org.enso.base.text.CaseFoldedString.Grapheme
 
 # needed by Comparator_Spec:
@@ -29,3 +29,5 @@ polyglot java import java.util.function.Function
 polyglot java import java.lang.Thread
 polyglot java import java.lang.Thread.State
 polyglot java import java.lang.Float
+
+CaseFoldedString = JCaseFoldedString
diff --git test/Base_Tests/src/Data/Text/Utils_Spec.enso test/Base_Tests/src/Data/Text/Utils_Spec.enso
index d0b987bea2..7d8d006ff7 100644
--- test/Base_Tests/src/Data/Text/Utils_Spec.enso
+++ test/Base_Tests/src/Data/Text/Utils_Spec.enso
@@ -1,7 +1,7 @@
 from Standard.Base import all
 
 polyglot java import org.enso.base.Text_Utils
-polyglot java import org.enso.base.text.CaseFoldedString
+import Standard.Base.Internal.Extra_Imports.CaseFoldedString
 polyglot java import com.ibm.icu.text.BreakIterator
 
 from Standard.Test import all
```
but that requires us to run with `--disable-private-check`. I can add such option, but then I am getting test failures at
```
[FAILED] Private constructors: [6/11, 26ms]
  | => enso / runEngineDistribution 111s
    - [FAILED] cannot directly call private constructor [4ms]
        Reason: Expected a Panic Private_Access to be thrown, but the action succeeded and returned [(Cons 42)] (at test/Base_Tests/src/Semantic/Private_Spec.enso:18:13-65).
    - [FAILED] can get reference to private constructor, but cannot call it [2ms]
        Reason: Expected a Panic Private_Access to be thrown, but the action succeeded and returned [(Cons 42)] (at test/Base_Tests/src/Semantic/Private_Spec.enso:27:13-58).
    - [FAILED] cannot get private field [2ms]
        Reason: Expected a Panic Private_Access to be thrown, but the action succeeded and returned [42] (at test/Base_Tests/src/Semantic/Private_Spec.enso:35:13-56).
    - [FAILED] cannot call private constructor from Java [3ms]
        Reason: Expected a Panic Private_Access to be thrown, but the action succeeded and returned [(Cons 42)] (at test/Base_Tests/src/Semantic/Private_Spec.enso:59:13-88).
    - [FAILED] cannot call private constructor from a lambda method [3ms]
        Reason: Expected a Panic Private_Access to be thrown, but the action succeeded and returned [(Cons 42)] (at test/Base_Tests/src/Semantic/Private_Spec.enso:62-64).

[FAILED] Private methods: [2/5, 9ms]
```
I guess we cannot have both: Run with `--disable-private-check` and these _Private tests_.

Commit:
Turn common-polyglot-core-utils into JPMS module

Commit:
Add depedencies to org.enso.std.base JPMS module

Commit:
Turn std-database into JPMS module

Commit:
Add depedencies to org.enso.std.aws JPMS module

JaroslavTulach commented on build.sbt:
```diff
@@ -3302,7 +3303,17 @@ lazy val `common-polyglot-core-utils` = project
       "org.graalvm.polyglot" % "polyglot" % graalMavenPackagesVersion % "provided"
-    )
+    ),
+    modulePath := {
```
Thanks you @Akirathan!

To turn a project into a module:
1. `enablePlugins(JPMSPlugin)`
2. define `modulePath` attribute with this magic
3. obviously add `module-info.java` into the project

What does it mean `filterModulesFromUpdate` anyway? Which _update_? What does `Compile / update` even mean? Why do we have to specify the module again when it is already among `libraryDependencies`?

Cannot by default `modulePath` be derived from _class path_ by filtering all JARs that have `module-info.class`?

Akirathan replied:
> What does it mean filterModulesFromUpdate anyway? Which update? What does Compile / update even mean?

`Compile / update` return an [UpdateReport](https://github.com/sbt/librarymanagement/blob/8d71f66997c356790fc887b17e33c7be66a4e590/core/src/main/contraband-scala/sbt/librarymanagement/UpdateReport.scala) that contains all the information about dependency resolution. `filterModulesFromUpdate` tries to pick the absolute path to the modules that are inside this report. This is the easiest way to do this functionality in sbt that I have been able to find.

> Why do we have to specify the module again when it is already among libraryDependencies?

Because of some technical limitations of sbt. `libraryDependencies` is of type `Seq[ModuleID]`. But `modulePath` needs to get `Seq[File]`. And `ModuleID` is nothing more than an Ivy package specification - a mere string. `JPMSUtils.filterModulesFromUpdate` is one way how to get a `File` from `ModuleID`.

> Cannot by default modulePath be derived from class path by filtering all JARs that have module-info.class?

In theory yes. I guess everything is possible with sbt. But in practice, I have no idea how to pull that off with sbt. I was hoping that in this stage, when we are defining more and more JPMS modules ourselves, we would be at least partially migrated to Bazel and that would help with it a bit.

Akirathan replied:
Further progress tracked in https://github.com/enso-org/enso/issues/10726

JaroslavTulach commented on build.sbt:
```diff
@@ -3518,6 +3519,25 @@ lazy val `std-database` = project
       "org.postgresql"       % "postgresql"              % "42.4.0"
     ),
+    modulePath := {
+      val externalModIds = libraryDependencies.value ++ Seq(
```
@Akirathan, I have no idea what this means or why such a complex setup for `modulePath` is necessary in the case of `std-database` project.

Akirathan replied:
This is a current technical limitation. Improvement tracked in https://github.com/enso-org/enso/issues/10726.

JaroslavTulach commented on build.sbt:
```diff
@@ -3575,6 +3576,31 @@ lazy val `std-aws` = project
       "software.amazon.awssdk" % "ssooidc"                 % awsJavaSdkV2Version
     ),
+    modulePath := {
+      val externalModIds = Seq(
```
@Akirathan, why do we have to repeat all these modules when they are already in `libraryDependencies`?

Akirathan replied:
This is a current technical limitation. Improvement tracked in https://github.com/enso-org/enso/issues/10726.

JaroslavTulach replied:
It is really simpler: cfdda7b

JaroslavTulach commented on build.sbt:
```diff
@@ -3575,6 +3576,31 @@ lazy val `std-aws` = project
+      )
+      val ourMods =
+        (`std-base` / Compile / exportedProducts).value.map(_.data) ++
+        (`std-database` / Compile / exportedProducts).value.map(_.data)
```
@Akirathan, cannot the `JPMSPlugin` iterate all project we `dependsOn` and extract modules from them automatically?

Akirathan replied:
This is a current technical limitation. Improvement tracked in https://github.com/enso-org/enso/issues/10726.

JaroslavTulach replied:
It is really simpler: cfdda7b

JaroslavTulach committed:
Using module-info to register SPI services

JaroslavTulach commented on std-bits/base/src/main/java/module-info.java:
```diff
@@ -1,5 +1,13 @@
+import org.enso.base.file_format.JSONFormatSPI;
+import org.enso.base.file_format.TextFormatSPI;
+import org.enso.base.file_format.XMLFormatSPI;
 import org.enso.base.file_system.FileSystemSPI;
```
@radeusgd, there is a small _terminology issue_ and encapsulation/security issue. [SPI stands for service provider interface](http://wiki.apidesign.org/wiki/SPI) - e.g. interface towards providers of some service. `FileFormatSPI` is such an interface. However `ByteFormatSPI`, `TextFormatSPI`, etc. _are not provider interfaces_!

Those are implementations. Thus adding `SPI` suffix to them isn't correct. I'd give them `Impl` suffix. Moreover such _implementations shouldn't be in a exported package_. Modules should export only an absolute minimum to [lower the conceptual surface](http://wiki.apidesign.org/wiki/Conceptual_surface) of the API. 

Some refactoring is desirable.

JaroslavTulach committed:
Removing another no longer needed dependency on NetBeans Lookup API

JaroslavTulach committed:
Removing add to class path concept

Commit:
Turn std-table into JPMS module

JaroslavTulach committed:
Move tests that need internal access to Standard.Base into test/Base_Internal_Tests

radeusgd commented on build/build/src/enso.rs:
```diff
@@ -100,6 +100,9 @@ impl BuiltEnso {
             // This flag enables assertions in the JVM. Some of our stdlib tests had in the past
             // failed on Graal/Truffle assertions, so we want to have them triggered.
             .set_env(JAVA_OPTS, &ide_ci::programs::java::Option::EnableAssertions.as_ref())?;
+        if test_path.as_str().contains("_Internal_") {
+            command.arg("--disable-private-check");
+        }
```
I like this approach :)

Ideally it would be nice if this was configurable through `package.yaml` (totally not a thing in scope of this PR obviously). But in the current context this check seems like a good heuristic.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Extra_Imports.enso:
```diff
@@ -26,6 +25,21 @@ polyglot java import java.math.BigInteger
+polyglot java import org.enso.base.CompareException as JCompareException
+polyglot java import org.enso.base.ObjectComparator as JObjectComparator
+
+Stream_Utils=JStream_Utils
+File_Utils=JFile_Utils
+CompareException=JCompareException
+ObjectComparator=JObjectComparator
+CaseFoldedString=JCaseFoldedString
+Text_Utils=JText_Utils
+BreakIterator=JBreakIterator
+Exception=JException
```
Don't we have to add `## PRIVATE` comments to all of these?

We definitely do not want them to show up in GUI's Component Browser.

JaroslavTulach replied:
The whole module is `private`. That was also the reason why I needed to `--disable-private-check`.

radeusgd replied:
Oh, I missed that 🤦

JaroslavTulach committed:
Resolving merge conficts

radeusgd commented on test/Base_Internal_Tests/src/Input_Output_Spec.enso:
```diff
@@ -0,0 +1,86 @@
+
+            is.available . should_equal 0
+
+        group_builder.specify "File in Dictionary supports host objects as keys" <|
+            # java.nio.path.Path has proper implementation of hashCode
+            dict = Dictionary.singleton (File_Utils.toPath "/home/user/file.txt") 42
+            dict.get "X" . should_equal Nothing
+            dict.get "A" . should_equal Nothing
+            dict.get (File_Utils.toPath "/home/user/file.txt") . should_equal 42
+
+        group_builder.specify "File equals on host values" <|
+            path1 = File_Utils.toPath "home" . resolve "user" . resolve "file.txt"
+            path2 = File_Utils.toPath "home" . resolve "user" . resolve "file.txt"
+            (path1 == path2).should_be_true
+            path3 = path1.resolve "subfile.txt"
+            (path3 == path2).should_be_false
```
These 2 feel like they belong in a separate group (or even file) from the ones above them.

Can we have a group e.g. `Host File values behaviour in Enso` or something?

JaroslavTulach replied:
Done in ae64b8acd4

JaroslavTulach replied:
Also ported to https://github.com/enso-org/enso/pull/10743/commits/4b0dd1110c12bc378e9489d342684e72b76aea49

JaroslavTulach committed:
No more of Polyglot.Java

JaroslavTulach committed:
Nicer formatting with var

JaroslavTulach committed:
Allow Standard.Table to depend on modules from Standard.Base

JaroslavTulach commented on distribution/lib/Standard/Table/0.0.0-dev/package.yaml:
```diff
@@ -8,3 +8,5 @@ authors:
   - name: Enso Team
     email: contact@enso.org
+requires:
+  - Standard.Base
```
By specifying `requires` on another package, the package _"sees modules"_ and can `require` them in own `module-info.java` files.

Akirathan replied:
How does this relate to the *project-private* and *friend* concept? Table Java code depends on modules from Base Java code, that is fine. But what about Table's Enso code and private stuff in Base? As far as I understand, all the classes inside `polyglot` directory of a project are considered project-private.

My question is: is this `requires` directive the same as *be friends with* directive? If so, should we rename to make it more clear?

JaroslavTulach replied:
The terminology is directly derived from the `module-info` terminology. `requires` is the same as `requires`. In Enso terms:

- if an Enso library Java module wants to use Java module from another Enso Library, it needs to `requires` it.

JaroslavTulach replied:
The biggest problem right now are the multiple `ModuleLayer`s. I am starting to get a feeling that while different modules may use other modules from _parent layer_, it is not possible to use `ServiceLoader` to load services across different _layers_. I still need to verify it, but it seems very likely it is the case.

JaroslavTulach replied:
I tried to do some research and write a sample project. Looks like I am not able to register `java.base` into my own layer - other than `ModuleLayer.boot()`. My code:
```java
package org.enso.test.layer.use;

import java.lang.module.Configuration;
import java.lang.module.ModuleFinder;
import java.util.Collections;
import java.util.List;

public class Use {

    public static void main(String[] args) {
        ModuleFinder finder = ModuleFinder.ofSystem();
        // final List<String> moduleNames = finder.findAll().stream().map(x -> x.descriptor().name()).collect(Collectors.toList());
        final List<String> moduleNames = Collections.singletonList("java.base");
        ModuleLayer emptyL = ModuleLayer.empty();
        ClassLoader emptyLoader = null; // ClassLoader.getSystemClassLoader();
        final List<Configuration> pConfs = Collections.singletonList(emptyL.configuration());
        Configuration pConf = Configuration.resolve(finder, pConfs, finder, moduleNames);
        List<ModuleLayer> pLayers = Collections.singletonList(emptyL);
        ModuleLayer.Controller cntrl = ModuleLayer.defineModulesWithOneLoader(pConf, pLayers, emptyLoader);
        System.out.println("Hello World!" + cntrl.layer().modules());
    }
}
```
opinions, @Akirathan?

Akirathan replied:
Look for an inspiration in [BenchProcessor.createTruffleModuleLayer](https://github.com/enso-org/enso/blob/develop/lib/scala/bench-processor/src/main/java/org/enso/benchmarks/processor/BenchProcessor.java#L53). I am creating ModuleLayer, and throwing away all the modules that are already present in boot layer there.

JaroslavTulach committed:
language-server/test compiles again

JaroslavTulach commented:
[Case study](https://github.com/JaroslavTulach/multilayerstudy/blob/8fb71ce75914865c68f3b057086342de8ffd03d1/README.md) shows how to organize layers properly for `ServiceLoader` to work.

JaroslavTulach committed:
Host Files values group_builder

JaroslavTulach committed:
Merging in develop branch

JaroslavTulach committed:
Merge with most recent develop branch that includes many JPMS

JaroslavTulach committed:
Using moduleDependencies and internalModuleDependencies for standard libraries

JaroslavTulach committed:
There can be Enso sources without associated package

JaroslavTulach commented:
The next [test failure to solve](https://github.com/enso-org/enso/actions/runs/11052195767/job/30703737248?pr=10714#step:7:8668) - `Module org.enso.std.base reads more than one module named com.ibm.icu`

Akirathan commented on engine/runtime-fat-jar/src/main/java/module-info.java:
This seems like a remnant after merging with develop. On develop, the whole `runtime-fat-jar` directory was deleted.

JaroslavTulach committed:
Merging with develop

JaroslavTulach committed:
Avoid duplicated modules. Use the one from boot layer.

JaroslavTulach committed:
Load all library Java modules with a single class loader

JaroslavTulach committed:
Opening std-table.jar for reflection from Enso code

JaroslavTulach committed:
Visualization_Tests run and yield just nine errors

JaroslavTulach committed:
Standard.Test loads classes from Standard.Base

JaroslavTulach committed:
Always prefer modules from parent layers

JaroslavTulach committed:
AWS requires access to Standard.Base and Standard.Database JARs

JaroslavTulach commented on distribution/lib/Standard/AWS/0.0.0-dev/package.yaml:
```diff
@@ -8,3 +8,6 @@ authors:
   - name: Enso Team
     email: contact@enso.org
+requires:
+  - Standard.Base
```
I am running into issues.
- if `Standard.AWS` has its own **isolated layer**, then it complains that it cannot find `org.enso.base.file_system.FileSystemSPI`
- but if I connect `Standard.AWS` layer with `Standard.Base`, then it complains _Jackson_ is there twice!

It is hard to have both - _isolation and connection_ between the layers.

Addressed by 405e88de540cfa6ee9c7254724f853825e96bf50

JaroslavTulach commented on distribution/lib/Standard/AWS/0.0.0-dev/package.yaml:
```diff
@@ -8,3 +8,6 @@ authors:
     email: contact@enso.org
+requires:
+  - Standard.Base
+  - Standard.Database
```
With 57e64e0d69fc332a93664593d765c35ba9ba10b5 I am getting the following when running `test/AWS_Tests`
```
Execution finished with an error: java.lang.module.ResolutionException: Modules aws.java.sdk.sts and aws.java.sdk.redshift export package com.amazonaws.auth.policy.actions to module software.amazon.awssdk.metrics
        at <java> java.base/java.lang.module.Resolver.resolveFail(Resolver.java:900)
        at <java> java.base/java.lang.module.Resolver.failTwoSuppliers(Resolver.java:814)
        at <java> java.base/java.lang.module.Resolver.checkExportSuppliers(Resolver.java:735)
        at <java> java.base/java.lang.module.Resolver.finish(Resolver.java:380)
        at <java> java.base/java.lang.module.Configuration.<init>(Configuration.java:139)
        at <java> java.base/java.lang.module.Configuration.resolveAndBind(Configuration.java:493)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.EnsoClassPath.create(EnsoClassPath.java:62)
```
We are not the only ones seeing this problem: https://github.com/aws/aws-sdk-java/issues/1658#issuecomment-430932125

JaroslavTulach replied:
Solved by creating our own [aws-wrapper](https://github.com/enso-org/enso/pull/10714/files#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91R1536). Thanks @Akirathan for your help!

JaroslavTulach committed:
Opening up std-table JAR, so Enso can access all Java classes

JaroslavTulach committed:
Less functional boilerplate

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/ClassPath7082

JaroslavTulach committed:
Special Lookup class and registration via a property to allow searching all the available ModuleLayers

JaroslavTulach committed:
Turning Standard.Image into Java module

JaroslavTulach committed:
Making test/Image_Tests pass OK

JaroslavTulach commented:
Currently solving problem with JDBC drivers:
```
sbt:enso> runEngineDistribution --run test/Table_Tests should.gracefully.handle.columns.from.different.backends
[FAILED] [SQLite In-Memory] (Core_Spec) Table.set: [0/1, 23ms]

    - [FAILED] should gracefully handle columns from different backends [23ms]
        Reason: Expected error Integrity_Error, but error There was an SQL error: No suitable driver found for jdbc:sqlite::memory:. has been returned (at test/Table_Tests/src/Common_Table_Operations/Core_Spec.enso:192:26-60).
```
The problem is that `DriverManager` does some fancy checks with `Reflection.getCallerClass()` and fails because `std-base` `ClassLoader` doesn't see drivers from `std-database` (that's what we wanted!).

Fixed by 3194f6d0bb

JaroslavTulach committed:
Avoid DriverManager. Directly talk to Driver instances.

JaroslavTulach committed:
Register SQLServerConnectionDetailsSPI in std-bits/microsoft/*/module-info.java

JaroslavTulach committed:
No polyglot symbol for org.enso.base.encoding.NewlineDetector in Table_Tests fix

JaroslavTulach committed:
No polyglot symbol for org.enso.table.data.table.Table in Table_Tests

JaroslavTulach committed:
No polyglot symbol for org.enso.table.data.table.join.between.SortedListIndex reported in Table_Tests

JaroslavTulach committed:
Using for loop

JaroslavTulach committed:
Make enso-test-java-helpers available to Table_Tests

JaroslavTulach commented:
Most of the `Table_Tests` are fixed by [today's commits]https://github.com/enso-org/enso/pull/10714/commits/6992103b2081621e703824d07da4a9284660e2e3). There are now three failing tests:
```bash
sbt:enso> runEngineDistribution --run test/Table_Tests should.recognise.a.SQLite.database.file|should.recognise.a.sqlite.file.by.extension.for.writing|should.connect.to.a.db.file
[FAILED] SQLite_Format should allow connecting to SQLite files: [0/3, 287ms]

    - [FAILED] should recognise a SQLite database file [166ms]
        Reason: Expected a value of type Standard.Database.Connection.SQLite_Format.SQLite_Format but got a value of type Standard.Base.Nothing.Nothing instead (at /home/devel/NetBeansProjects/enso/enso.ide/test/Table_Tests/src/Database/SQLite_Spec.enso:422:13-80).


    - [FAILED] should recognise a sqlite file by extension for writing [52ms]
        Reason: Expected a value of type Standard.Database.Connection.SQLite_Format.SQLite_Format but got a value of type Standard.Base.Nothing.Nothing instead (at /home/devel/NetBeansProjects/enso/enso.ide/test/Table_Tests/src/Database/SQLite_Spec.enso:425:13-114).


    - [FAILED] should connect to a db file [68ms]
        Reason: An unexpected dataflow error ((Unsupported_Type (File_Format_Metadata.Value '/home/devel/NetBeansProjects/enso/enso.ide/test/Table_Tests/data/transient/sqlite_test.db' 'sqlite_test.db' '.db' File.read_first_bytes[File.enso:732-734] self=org.enso.interpreter.runtime.data.EnsoFile@5248c05a n=_ Nothing))) has been matched (at /home/devel/NetBeansProjects/enso/enso.ide/test/Table_Tests/src/Database/SQLite_Spec.enso:450:13-50).
        at <enso> Auto_Detect.read<arg-1>(/home/devel/NetBeansProjects/enso/enso.ide/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Base/0.0.0-dev/src/System/File_Format.enso:57:35-80)


0 tests succeeded.
3 tests failed.
0 tests skipped.
0 groups skipped.
```
When there is a resolution problem, report normal exception, don't crash the JVM:
```
Execution finished with an error: java.lang.module.FindException: Module org.enso.std.table not found, required by org.enso.test.java_helpers
        at <java> java.base/java.lang.module.Resolver.findFail(Resolver.java:892)
        at <java> java.base/java.lang.module.Resolver.resolve(Resolver.java:192)
        at <java> java.base/java.lang.module.Resolver.resolve(Resolver.java:141)
        at <java> java.base/java.lang.module.Configuration.resolveAndBind(Configuration.java:491)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.EnsoClassPath.create(EnsoClassPath.java:58)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.EnsoContext.findClassPath(EnsoContext.java:497)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.EnsoContext.lookupJavaClass(EnsoContext.java:582)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.IrToTruffle.$anonfun$registerPolyglotImports$2(IrToTruffle.scala:247)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.util.CachingSupplier.get(CachingSupplier.java:48)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.IrToTruffle$ExpressionProcessor.processCaseBranch(IrToTruffle.scala:1761)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.IrToTruffle$ExpressionProcessor.$anonfun$processCase$1(IrToTruffle.scala:1421)
        at <java> scala.library@2.13.11/scala.collection.Iterator$$anon$9.next(Iterator.scala:584)
```

JaroslavTulach committed:
No need to generate .bindings index to run benchmarks

JaroslavTulach committed:
BenchProcessor puts runtime.jar into own layer. Use that one as bootLayer() when specified.

JaroslavTulach committed:
test/Benchmarks need access to org.enso.table.data.index.OrderedMultiValueKey

JaroslavTulach committed:
Introducing AWS wrapper module

JaroslavTulach committed:
Placeholders for exported packages

Commit:
Remove placeholders

Commit:
Remove aws-wrapper/Compile/moduleDependencies.

They are not required - we don't want anything on module-path

Commit:
javaModuleName is the first setting

Commit:
Fix typo in cbor library dependency

Commit:
Typo in module name.

The module name in module-info.java must be the same as `javaModuleName` sbt setting.

Commit:
Fix compilation of aws-wrapper

- Patch all AWS modules
- Fix assemblyExcludedJars to remove all transitive dependencies

Commit:
Remove empty settings from std-aws

Commit:
Fix typo in module name

Commit:
aws-wrapper-assembly.jar gets into component directory.

Commit:
Remove signature from aws-wrapper-assembly jar

Commit:
aws-wrapper dependency of std-aws is "provided".

So there is just a single jar in AWS/polyglot/java directory

Commit:
Ensure aws-wrapper-assembly.jar gets into AWS/polyglot/java directory

Commit:
Fix NoSuchElementException in JPMSUtils

Commit:
Revert "aws-wrapper-assembly.jar gets into component directory."

This reverts commit c97ec776ffd6f43f01f0b7dd35f14db91d12a166.

Commit:
org.enso.std.aws exports org.enso.aws package

Commit:
slf4j is module dependency of aws-wrapper

Commit:
aws-wrapper depends on org.reactivestreams

Commit:
aws-wrapper uses its own service.

Fixes ServiceConfigurationError.

Commit:
Add some dependencies to aws-wrapper

Commit:
Add necessary exports

Commit:
fmt

Commit:
FileSystemSPI uses Lookup instead of ServiceLoader

Commit:
org.enso.std.aws provides DataLinkSPI.

And DataLinkSPI is implemented with Lookup instead of ServiceLoader.

Commit:
Better error message on duplicated service providers

Commit:
Duplicate provider is just a warning

JaroslavTulach committed:
List is easier to == than Array

JaroslavTulach committed:
runtime-integration-tests are loading classes with empty EnsoClassPath

JaroslavTulach committed:
Java.lookup_class has been removed

JaroslavTulach committed:
Using List() at the right places

JaroslavTulach commented:
Heuréka. [JVM tests](https://github.com/enso-org/enso/actions/runs/11228376996/job/31212158218?pr=10714) are green.

JaroslavTulach committed:
std.table module isn't present in NI enso build

JaroslavTulach committed:
Use Lookup to lookup DatabaseConnectionDetailsSPI

JaroslavTulach committed:
An unexpected panic was thrown: (No_Such_Method.Error org.enso.base.net.http.UrlencodedBodyBuilder UnresolvedSymbol<new>)

JaroslavTulach committed:
Initialize ModuleLayer for a library sooner than a class is loaded from it

JaroslavTulach committed:
There is no runtime-fat-jar anymore

Commit:
Add JDK_MODULES to small JDK of engine-runner native image

JaroslavTulach committed:
Catch also ResolutionException

JaroslavTulach committed:
Always add at least boot layer as parent

JaroslavTulach committed:
Make Tableau_Tests pass

JaroslavTulach commented on build.sbt:
```diff
@@ -3633,7 +3633,7 @@ lazy val `engine-runner` = project
         "org.graalvm.nativeimage,org.graalvm.nativeimage.builder,org.graalvm.nativeimage.base,org.graalvm.nativeimage.driver,org.graalvm.nativeimage.librarysupport,org.graalvm.nativeimage.objectfile,org.graalvm.nativeimage.pointsto,com.oracle.graal.graal_enterprise,com.oracle.svm.svm_enterprise"
       val JDK_MODULES =
-        "jdk.localedata,jdk.httpserver,java.naming,java.net.http"
+        "jdk.localedata,jdk.httpserver,java.naming,java.net.http,java.desktop,java.xml.crypto,jdk.xml.dom,java.sql.rowset"
```
This will very likely break Espresso support. We had to remove `java.desktop` because of Espresso in #9866

Akirathan replied:
`java.desktop` is required by `org.apache.poi.poi`. The NI failed on `Base_Tests` because of that.

JaroslavTulach replied:
> `java.desktop` is required by `org.apache.poi.poi`. The NI failed on `Base_Tests` because of that.

That's scary. Probably another dependency on `java.beans` - we should change upstream to make this dependency optional.

JaroslavTulach replied:
As a result of using `java.desktop` the _native image_ build now adds:
```
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libawt.so (jdk_library)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libawt_headless.so (jdk_library)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libawt_xawt.so (jdk_library)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libfontmanager.so (jdk_library)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libjava.so (jdk_library_shim)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libjavajpeg.so (jdk_library)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/libjvm.so (jdk_library_shim)
[info]  /built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/liblcms.so (jdk_library)
```
that's unfortunate.

JaroslavTulach replied:
Backed out in 4c46657

JaroslavTulach committed:
Bringing up-to-date with most recent develop

JaroslavTulach committed:
SQLServerDataLinkSPI is registered in module-info

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/ClassPath7082

JaroslavTulach commented:
There are seven `[FAILED]` tests when running the _standard library tests_ on Linux:
- [SQLServer](https://github.com/enso-org/enso/actions/runs/11434005820/job/31806847233?pr=10714#step:7:2584) fails with `An unexpected dataflow error ((Missing_Data_Link_Library.Error 'Standard.Microsoft' 'SQLServer_Connection'))` 
- `Table_Tests` fail on [PostgreSQL](https://github.com/enso-org/enso/actions/runs/11434005820/job/31806847233?pr=10714#step:7:3383) with `An unexpected dataflow error (Not_Found)` probably not founding the driver

I need to find a way to simulate the problem locally!

**Update**: I manage to simulate similar problem with 208d05d:
```
sbt:enso> engine-runner/buildNativeImage
sbt:enso> runEngineDistribution --run test/Base_Tests/src/Data/XML/XML_Spec.enso Can.read.from.a.file
- [FAILED] Can read from a file [26ms]
        Reason: An unexpected dataflow error ((Unsupported_Type (File_Format_Metadata.Value '/Base_Tests/data/xml/sample.xml' 'sample.xml' '.xml' File.read_first_bytes[File.enso:732-734] self=org.enso.interpreter.runtime.data.EnsoFile@1e09a7bb n=_ Nothing))) h
```
looks like there is a problem in discovery of `XMLFormatSPI` somewhere in [Lookup](https://github.com/enso-org/enso/pull/10714/files#diff-ce283a1bbc5cd6c0dd0b2a1d6da639497a525aeb40bef2c4ed236e2fd10e64d6R61) class.

radeusgd commented:
> * `Table_Tests` fail on [PostgreSQL](https://github.com/enso-org/enso/actions/runs/11434005820/job/31806847233?pr=10714#step:7:3383) with `An unexpected dataflow error (Not_Found)` probably not founding the driver

This `Not_Found` is not about the driver. It looks like #11278. So theoretically merging `develop` should help, but it looks like you have already done this. So that's weird.

JaroslavTulach committed:
Backing out ec03e1080fbe19f7b7363206e876e72298bbc0f2

JaroslavTulach committed:
Include chrome and DAP debugging tools in the enso executable

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/ClassPath7082

JaroslavTulach committed:
uses/provides instead of @ServiceProvider

JaroslavTulach commented:
I tried to workaround the GraalVM `ServiceLoader` bugs (like https://github.com/oracle/graal/issues/9952#issuecomment-2442086616) with:
```diff
diff --git build.sbt build.sbt
index 101f03aa16..f3d1e4a9c9 100644
--- build.sbt
+++ build.sbt
@@ -3777,7 +3777,6 @@ lazy val `engine-runner` = project
               "com.sun.jna",
               "com.microsoft",
               "akka.http",
-              "org.enso.base",
               "org.enso.image",
               "org.enso.table"
             )
diff --git engine/runner/src/main/java/org/enso/runner/EnsoLibraryFeature.java engine/runner/src/main/java/org/enso/runner/EnsoLibraryFeature.java
index 433271da46..2d17c38fcb 100644
--- engine/runner/src/main/java/org/enso/runner/EnsoLibraryFeature.java
+++ engine/runner/src/main/java/org/enso/runner/EnsoLibraryFeature.java
@@ -16,6 +16,34 @@ import org.graalvm.nativeimage.hosted.RuntimeReflection;
 public final class EnsoLibraryFeature implements Feature {
   @Override
   public void beforeAnalysis(BeforeAnalysisAccess access) {
+    try {
+      var lookupClass = access.findClassByName("org.enso.base.lookup.Lookup");
+      var foundField = lookupClass.getDeclaredField("found");
+      foundField.setAccessible(true);
+      var reloadMethod = lookupClass.getDeclaredMethod("reload");
+      access.registerFieldValueTransformer(
+          foundField,
+          (receiver, originalValue) -> {
+            try {
+              System.err.println("  transform " + receiver);
+              reloadMethod.invoke(receiver);
+              var newValue = foundField.get(receiver);
+              System.err.println("  value " + newValue);
+              return newValue;
+            } catch (ReflectiveOperationException ex) {
+              throw new IllegalStateException(ex);
+            }
+          });
+
+    } catch (ReflectiveOperationException | SecurityException ex) {
+      throw new IllegalStateException(ex);
+    }
+
+    registerPolyglotJavaImportForReflectiveAccess(access);
+  }
+
+  private void registerPolyglotJavaImportForReflectiveAccess(BeforeAnalysisAccess access)
+      throws IllegalStateException {
     var libs = new LinkedHashSet<Path>();
     for (var p : access.getApplicationClassPath()) {
       var p1 = p.getParent();
@@ -30,18 +58,18 @@ public final class EnsoLibraryFeature implements Feature {
     }
 
     /*
-      To run Standard.Test one shall analyze its polyglot/java files. But there are none
-      to include on classpath as necessary test classes are included in Standard.Base!
-      We can locate the Test library by following code or we can make sure all necessary
-      imports are already mentioned in Standard.Base itself.
+    To run Standard.Test one shall analyze its polyglot/java files. But there are none
+    to include on classpath as necessary test classes are included in Standard.Base!
+    We can locate the Test library by following code or we can make sure all necessary
+    imports are already mentioned in Standard.Base itself.
 
     if (!libs.isEmpty()) {
-      var f = libs.iterator().next();
-      var stdTest = f.getParent().getParent().resolve("Test").resolve(f.getFileName());
-      if (stdTest.toFile().exists()) {
-        libs.add(stdTest);
-      }
-      System.err.println("Testing library: " + stdTest);
+    var f = libs.iterator().next();
+    var stdTest = f.getParent().getParent().resolve("Test").resolve(f.getFileName());
+    if (stdTest.toFile().exists()) {
+    libs.add(stdTest);
+    }
+    System.err.println("Testing library: " + stdTest);
     }
     */
 
diff --git engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoClassPath.java engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoClassPath.java
index 2d249e3401..bf9a7f11ea 100644
--- engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoClassPath.java
+++ engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoClassPath.java
@@ -120,6 +120,7 @@ public final class EnsoClassPath {
 
   @SuppressWarnings("unchecked")
   private static void registerLayer(ModuleLayer moduleLayer) {
+    new Exception("registerLayer").printStackTrace();
     var props = System.getProperties();
     Collection<ModuleLayer> layers;
     if (props.get("enso.class.path") instanceof Collection registeredLayers) {
diff --git std-bits/base/src/main/java/org/enso/base/lookup/Lookup.java std-bits/base/src/main/java/org/enso/base/lookup/Lookup.java
index 439bf03f22..f42b182d27 100644
--- std-bits/base/src/main/java/org/enso/base/lookup/Lookup.java
+++ std-bits/base/src/main/java/org/enso/base/lookup/Lookup.java
@@ -33,7 +33,7 @@ public final class Lookup<T> implements Iterable<T> {
   }
 
   public void reload() {
-    found = null;
+    found = findAll();
   }
 
   private List<ServiceLoader.Provider<T>> findAll() {
```

The idea is to populate all the `ServiceLoader` instances during _image build time_.  There is `FieldValueTransformer` trying to locate all services. However that doesn't work as the **list of packages isn't yet registered**!

JaroslavTulach commented:
This PR https://github.com/oracle/graal/pull/10202 shows the substitutions needed to make the `ServiceLoader` work in modular environment.

>  _Native Image_ support for `ServiceLoader` is [implemented](https://github.com/oracle/graal/issues/9952#issuecomment-2497642182) and will be part of next GraalVM

- #11868 might be a better alternative to overcome the problems of `ServiceLoader`
- because it also solves #11707
- we don't need https://github.com/JaroslavTulach/multilayerstudy tricks to make `ServiceLoader` work with multiple layers

JaroslavTulach commented:
To progress, we need:
- #12452
- because of https://github.com/oracle/graal/issues/9952 bugfix

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoContext.java:
```diff
@@ -82,7 +85,7 @@ public final class EnsoContext {
   private final EnsoLanguage language;
   private final Env environment;
-  private final HostClassLoader hostClassLoader = new HostClassLoader();
+  private final Map<Package<?>, EnsoClassPath> classPaths = new HashMap<>();
```
Attempts to provide independent classloader for each library

JaroslavTulach committed:
Somehow merging with latest develop. Had to drop all std-bits changes, but engine/runtime compiles fine

JaroslavTulach commented:
- I tried to merge this PR with most recent `develop`
- but there is so many conflicts (especially in `std-bits`)
- that it is better to start from scratch
- or pick just pieces of this PR up
